chain-vote/src/encryption.rs

```diff
@@ -0,0 +1,385 @@
+#![allow(dead_code)]
+
+//! Implementation of the different encryption/decryption mechanisms used in `chain-vote`, including their
+//! corresponding structures. In particular, we use (lifted) ElGamal cryptosystem, and combine with ChaCha
+//! stream cipher to produce a hybrid encryption scheme.
+
+use crate::gang::{GroupElement, Scalar};
+use rand_core::{CryptoRng, RngCore};
+use std::ops::{Add, Mul};
+
+use cryptoxide::blake2b::Blake2b;
+use cryptoxide::chacha20::ChaCha20;
+use cryptoxide::digest::Digest;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+/// ElGamal public key. pk = sk * G, where sk is the `SecretKey` and G is the group
+/// generator.
+pub struct PublicKey {
+    pub pk: GroupElement,
+}
+
+#[derive(Clone)]
+/// ElGamal secret key
+pub struct SecretKey {
+    pub sk: Scalar,
+}
+
+#[derive(Clone)]
+/// ElGamal keypair
+pub struct Keypair {
+    pub secret_key: SecretKey,
+    pub public_key: PublicKey,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// ElGamal ciphertext. Given a message M represented by a group element, and ElGamal
+/// ciphertext consists of (r * G; M + r * `PublicKey`), where r is a random `Scalar`.
```

Comment:
Not really. A group has a binary operation, with certain properties. It can be represented with additive or multiplicative notation, without necessarily committing to the fact that the operation is addition or multiplication. If the binary operation is abelian, then + is usually preferred. Given that in our context we are using groups over elliptic curves, the operation is abelian, and, in particular, is the sum of two points. Hence my preference in using additive notation.